conda_forge_tick/utils.py

```diff
@@ -315,6 +315,62 @@ def parse_recipe_yaml_containerized(
     )
 
 
+def _flatten_requirement_pin_dicts(
+    recipes: list["RecipeTypedDict"],
+) -> list["RecipeTypedDict"]:
+    """
+    Flatten out any "pin" dictionaries.
+
+    The `${{ pin_subpackage(foo) }}` and `${{ pin_compatible(foo) }}` functions
+    are converted into dictionaries. However, we want to extract the name only.
+    """
+
+    def flatten_pin(s: str | dict[str, Any]) -> str:
+        if isinstance(s, dict):
```

Comment:
We probably want to use a class like `MutableMapping` here so that we cover cases where non-standard dict-like containers are passed in.